### ISSUE TITLE
type.isTuple is false for destructured tuples

### ISSUE THREAD

**Issue by Gerrit0** [dsherret/ts-simple-ast#519]
Version: 20.0.0

Types seem to break when using `...` in destructured declarations, I would expect `isTuple` to be true, since hovering over `rest` in the following code shows `[number]`, but both `isTuple` and `isArray` return false.

```ts
const [ a, ...rest ] = [1, 2];
export { rest };
```

**To Reproduce**

```ts
import { Project, ts } from "ts-simple-ast";

console.log(ts.version); // => 3.2.2

const project = new Project();

const sourceFile = project.createSourceFile(
    "test.ts",
    `
const [ a, ...rest ] = [1, 2];
export { rest };
`
);

const [ rest ] = sourceFile.getExportSymbols(); // Not sure how else to get to rest
const type = rest.getDeclaredType();

console.log('Tuple', type.isTuple()); // => false
console.log('Array', type.isArray()); // => false

```

**Expected behavior**

`type.isTuple()` should return true.

**Comment by Gerrit0** [comment]
After looking in to this a bit more, I'm not entirely sure it is a bug. In this case the type is a reference type... which points to a tuple type (that does have `ts.ObjectFlags.Tuple` set)

The current behavior is unexpected if you aren't familiar with the compiler (me, most people), but I'm not sure it is *wrong*.

**Comment by dsherret** [comment]
@Gerrit0 Thanks for the update! I didn't investigate this, but I was suspecting there was a chance it wasn't a bug. I'll look at it soon... I took a break during the holidays from working on open source, but I'll get back into it now that it's the new year.

**Comment by dsherret** [comment]
@Gerrit0 in that scenario, using `.getDeclaredType()` will return a type of `any` (`type.getText()`).

I found it works fine if I do this:

```ts
rest.getTypeAtLocation(rest.getDeclarations()[0]).isTuple(); // true
```

Or this (opened #521 to make it easier to get the identifier):

```ts
const restIdentifier = sourceFile.getDescendantsOfKind(SyntaxKind.Identifier).find(n => n.getText() === "rest")!;
const restType = restIdentifier.getType();

console.log(restType.isTuple()); // true
```

To be honest, I'm not sure what `.getDeclaredType()` is supposed to do, but I've always found `.getTypeAtLocation(node)` to work.

By the way, `isTuple()` will look at the reference type's target type so that should work if it's a reference type.

In conclusion, I'm thinking this is not a bug?

**Comment by Gerrit0** [comment]
I agree, this isn't actually a bug. #521 looks great, thanks! Closing issue.
